Race condition getting node candidates during parallel deployment

The exn-connector-java SyncedPublisher class is not thread-safe and does not support multiple requests; create one instance per request, register it before the request then unregister it afterwards.

Instructions copied from Slack discussion:

1. The library was not being used correctly, regarding the initial EventLoop
2. A SyncedPublisher **cannot** be used across threads (it is not thread safe in that sense, but it is also an antipattern). The syncedpublisher itself spawns its own **single** thread, a Publisher can be used, but you need to handle the asynchronous `onMessage`.
3. SyncedPublisher can be created in the onMessage or onReady callback, and destroyed
4. The EXN client itself spawns its own thread in the `start` method, and you can initiate everythin in the `onReady`
5. The context, contains the AtomicReferences of the publishers which you can get **from the context** to ensure all communication with the broker has actually been initialized. or re-initialized [...]

Here is the re-written version of the test case so that you can see some of the best practices when using the library

```java
package eu.nebulouscloud.exn.core;

import java.io.IOException;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.exc.StreamReadException;
import com.fasterxml.jackson.databind.DatabindException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import eu.nebulouscloud.exn.Connector;
import eu.nebulouscloud.exn.core.SyncedPublisher;
import eu.nebulouscloud.exn.handlers.ConnectorHandler;
import eu.nebulouscloud.exn.settings.StaticExnConfig;

/**
 * This test demonstrates an issue with parallel synced requests using EXN middleware library.
 * Using only one EXN Middleware connector, creates two threads that will issue requests to `eu.nebulouscloud.exn.sal.cluster.get`. 
 * One of the threads will request the status of an existing cluster, and thus, should get a valid response. 
 * The other requests the status of a cluster that is not existing, and thus, should get an error as response.
 * 
 * This test demonstrates that, at random, responses from both threads are crossed, leading to errors.
 *
 * 
 * This is a sample of the output:
 * 
 * This is a normal beheviour for T1:
[INFO ] 2024-10-15 13:09:40.312 [Thread-0] TestParallelGetCluster - T1->isClusterReady: {"clusterId":"8a7484f7928f404a01928fca93190698","name":"6d7a1-2","master-node":"m6d7a1-2-master",............"}
[INFO ] 2024-10-15 13:09:40.313 [Thread-0] TestParallelGetCluster - T1->OK


* This is a normal behaviour for T2:
[INFO ] 2024-10-15 13:09:40.312 [Thread-1] TestParallelGetCluster - T2->isClusterReady: NULL
[INFO ] 2024-10-15 13:09:40.313 [Thread-1] TestParallelGetCluster - T2->OK

* This is an abnormal behaviour for T1:
[ERROR] 2024-10-15 13:09:41.441 [Thread-0] TestParallelGetCluster - T1->exn-middleware-sal request failed with error code '' and message '', caller 'getCluster'
[INFO ] 2024-10-15 13:09:41.441 [Thread-0] TestParallelGetCluster - T1->isClusterReady: NULL
[ERROR] 2024-10-15 13:09:41.442 [Thread-0] TestParallelGetCluster - T1->Incorrect answer!!!

*This is an abnormal behaviour for T2:
[ERROR] 2024-10-15 13:09:46.098 [Thread-1] TestParallelGetCluster - T2->exn-middleware-sal request failed with error code '' and message '', caller 'getCluster'

 */

public class TestParallelGetCluster {

	//This variable should be set to a name of a cluster existing on Proactive
	//static final String VALID_CLUSTER_NAME = "1ed4a-3";
	static final String VALID_CLUSTER_NAME = "12005-7";
	static Logger LOGGER = LoggerFactory.getLogger(TestParallelGetCluster.class);


	private Connector myEXNClient;

	private class MyConnectorHandler extends ConnectorHandler {

		@Override
		public void onReady(Context context) {


			/**
			* Runnable that periodically issues requests to `eu.nebulouscloud.exn.sal.cluster.get` to get the status of the cluster `clusterName`.
			* For each requests, checks if the result retrieved from SAL is valid or not (contains a valid JSON with the property `status` in it).
			* Logs if the validity of the result is expected or not.
			 */


			LOGGER.info("Starting test");
			//This thread will request for an existing cluster, and thus, should always get valid responses
			Thread t1 = null, t2 = null;

			//You can comment/uncoment T1 and T2 alternatively to see the results.
			t1 = new Thread(new GetClusterRunnable("T1", VALID_CLUSTER_NAME, true, (SyncedPublisher) context.getPublisher("getCluster-"+VALID_CLUSTER_NAME) ));
			t2 = new Thread(new GetClusterRunnable("T2", "thisclusterforsurewontexist", false,(SyncedPublisher)  context.getPublisher("thisclusterforsurewontexist") ));
			if(t1!=null) t1.start();
			if(t2!=null) t2.start();
		}
	}

	private TestParallelGetCluster() throws InterruptedException {
		myEXNClient = new Connector(
				"thisINotImportant",
				new MyConnectorHandler(),
				List.of(
					new SyncedPublisher("getCluster-"+VALID_CLUSTER_NAME, "eu.nebulouscloud.exn.sal.cluster.get", true, true),
					new SyncedPublisher("thisclusterforsurewontexist", "eu.nebulouscloud.exn.sal.cluster.get", true, true)
				),
				List.of(),
				true,
				true, new StaticExnConfig("localhost", 5672, "admin", "admin"));
		myEXNClient.start();


	}



	static class GetClusterRunnable implements Runnable {

			String appID;
			String clusterName;
			boolean shouldBeValid = false;
			protected ObjectMapper om = new ObjectMapper();
			SyncedPublisher publisher;

			public GetClusterRunnable(String appID, String clusterName, boolean shouldBeValid, SyncedPublisher publisher) {
				this.appID = appID;
				this.clusterName = clusterName;
				this.shouldBeValid = shouldBeValid;
				this.publisher = publisher;
			}

			private JsonNode extractPayloadFromExnResponse(Map<String, Object> responseMessage, String caller) {
				JsonNode response = om.valueToTree(responseMessage);
				String salRawResponse = response.at("/body").asText();
				JsonNode metadata = response.at("/metaData");
				JsonNode salResponse = om.missingNode();
				try {
					salResponse = om.readTree(salRawResponse);
				} catch (JsonProcessingException e) {
					LOGGER.error("Could not read message body as JSON: body = '{}', caller = '{}'", salRawResponse,
							caller, e);
					return om.missingNode();
				}
				if (!metadata.at("/status").asText().startsWith("2")) {
					// we only accept 200, 202, numbers of that nature
					LOGGER.error(appID + "->exn-middleware-sal request failed with error code '{}' and message '{}', caller '{}'",
							metadata.at("/status"), salResponse.at("/message").asText(), caller);
					return om.missingNode();
				}
				return salResponse;
			}

			public String getClusterStatus(String clusterName) {
				Map<String, Object> msg = Map.of("metaData", Map.of("user", "admin", "clusterName", clusterName));
				Map<String, Object> response = this.publisher.sendSync(msg, clusterName, null, false);
				JsonNode payload = extractPayloadFromExnResponse(response, "getCluster");
				if (payload.isMissingNode())
				{
					LOGGER.info(appID + "->isClusterReady: NULL");
					return null;
				}
				LOGGER.info(appID + "->isClusterReady: " + payload.toString());
				JsonNode jsonState = payload.at("/status");
				if (jsonState.isMissingNode())
					return null;
				return jsonState.asText();
			}

			@Override
			public void run() {
				while (true) {
					try {
						String result = getClusterStatus(clusterName);
						if (result == null) {
							if (shouldBeValid)
								LOGGER.error(appID + "->" + "Incorrect answer!!! ");
							else
								LOGGER.info(appID + "->OK");
						} else {
							if (!shouldBeValid)
								LOGGER.error(appID + "->" + "Incorrect answer!!!");
							else
								LOGGER.info(appID + "->OK");
						}
					} catch (Exception e) {
						e.printStackTrace();
					}
					try {
						Thread.sleep(1000);
					} catch (InterruptedException e) {
						e.printStackTrace();
					}
				}

			}
		}
		;
	public static void main(String[] args)
			throws StreamReadException, DatabindException, IOException, InterruptedException {
		new TestParallelGetCluster();
	}
}
```

--- a/optimiser-controller/src/main/java/eu/nebulouscloud/optimiser/controller/ExnConnector.java
+++ b/optimiser-controller/src/main/java/eu/nebulouscloud/optimiser/controller/ExnConnector.java
@@ -27,6 +27,8 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
 
 /**
@@ -46,6 +48,32 @@ public class ExnConnector {
 
     /** The Connector used to talk with ActiveMQ */
     private final Connector conn;
+
+    private Context context_ = null;
+    /**
+     * Safely obtain the connection Context object.  Since the {@link
+     * #context_} field is set asynchronously after the ExnConnector
+     * constructor has finished, there is a race condition where we might hit
+     * a null value if using the field directly.
+     */
+    public Context getContext() {
+        if (context_ == null) {
+            synchronized(this) {
+                while (context_ == null) {
+                    try {
+			wait();
+		    } catch (InterruptedException e) {
+                        log.error("Caught InterruptException while waiting for ActiveMQ connection Context; looping", e);
+		    }
+                }
+            }
+        }
+        return context_;
+    }
+
+    /** A counter to create unique names for SyncedPublisher instances. */
+    private AtomicInteger publisherNameCounter = new AtomicInteger(1);
+
     /** if non-null, signals after the connector is stopped */
     private CountDownLatch synchronizer = null;
 
@@ -101,37 +129,6 @@ public class ExnConnector {
     @Getter
     private final Publisher appStatusPublisher;
 
-    // ----------------------------------------
-    // Communication with SAL
-
-    /** The findNodeCandidates endpoint.  Should not be used during normal
-      * operation--ask the broker instead. */
-    public final SyncedPublisher findSalNodeCandidates;
-    /** The findNodeCandidates endpoint (Broker's version).  This one adds
-      * attributes "score", "rank" to the answer it gets from SAL. */
-    public final SyncedPublisher findBrokerNodeCandidates;
-    /** The findNodeCandidatesMultiple endpoint offered by the cloud-fog
-      * service broker.  Like findBrokerNodeCandidates but we send multiple
-      * requirement lists that the broker then aggregates into a unified
-      * ranked node candidate list. */
-    public final SyncedPublisher findBrokerNodeCandidatesMultiple;
-    /** The defineCluster endpoint. */
-    public final SyncedPublisher defineCluster;
-    /** The getCluster endpoint. */
-    public final SyncedPublisher getCluster;
-    /** The labelNodes endpoint. */
-    public final SyncedPublisher labelNodes;
-    /** The deployCluster endpoint. */
-    public final SyncedPublisher deployCluster;
-    /** The deployApplication endpoint. */
-    public final SyncedPublisher deployApplication;
-    /** The scaleOut endpoint. */
-    public final SyncedPublisher scaleOut;
-    /** The scaleIn endpoint. */
-    public final SyncedPublisher scaleIn;
-    /** The deleteCluster endpoint. */
-    public final SyncedPublisher deleteCluster;
-
     /**
      * Create a connection to ActiveMQ via the exn middleware, and set up the
      * initial publishers and consumers.
@@ -140,45 +137,31 @@ public class ExnConnector {
      * @param port the port of the ActiveMQ server (usually 5672)
      * @param name the login name to use
      * @param password the login password to use
-     * @param callback A ConnectorHandler object.  Its {@link
-     *  ConnectorHandler#onReady} method will be called after the {@link
-     *  Connector#start} method has connected and set up all handlers.
      */
-    public ExnConnector(String host, int port, String name, String password, ConnectorHandler callback) {
+    public ExnConnector(String host, int port, String name, String password) {
         amplMessagePublisher = new Publisher("controller_ampl", ampl_message_channel, true, true);
         metricListPublisher = new Publisher("controller_metric_list", metric_list_channel, true, true);
         appStatusPublisher = new Publisher("app_status", app_status_channel, true, true);
-        findSalNodeCandidates = new SyncedPublisher("findSalNodeCandidates", "eu.nebulouscloud.exn.sal.nodecandidate.get", true, true);
-        findBrokerNodeCandidates = new SyncedPublisher("findBrokerNodeCandidates", "eu.nebulouscloud.cfsb.get_node_candidates", true, true);
-        findBrokerNodeCandidatesMultiple = new SyncedPublisher("findBrokerNodeCandidatesMultiple", "eu.nebulouscloud.cfsb.get_node_candidates_multi", true, true);
-        defineCluster = new SyncedPublisher("defineCluster", "eu.nebulouscloud.exn.sal.cluster.define", true, true);
-        getCluster = new SyncedPublisher("getCluster", "eu.nebulouscloud.exn.sal.cluster.get", true, true);
-        labelNodes = new SyncedPublisher("labelNodes", "eu.nebulouscloud.exn.sal.cluster.label", true, true);
-        deployCluster = new SyncedPublisher("deployCluster", "eu.nebulouscloud.exn.sal.cluster.deploy", true, true);
-        deployApplication = new SyncedPublisher("deployApplication", "eu.nebulouscloud.exn.sal.cluster.deployapplication", true, true);
-        scaleOut = new SyncedPublisher("scaleOut", "eu.nebulouscloud.exn.sal.cluster.scaleout", true, true);
-        scaleIn = new SyncedPublisher("scaleIn", "eu.nebulouscloud.exn.sal.cluster.scalein", true, true);
-        deleteCluster = new SyncedPublisher("deleteCluster", "eu.nebulouscloud.exn.sal.cluster.delete", true, true);
 
-        conn = new Connector("optimiser_controller",
-            callback,
+        conn = new Connector(
+            "optimiser_controller",
+            new ConnectorHandler() {
+                public void onReady(AtomicReference<Context> context) {
+                    synchronized(ExnConnector.this) {
+                        // Make sure no one tries to use a null context
+                        ExnConnector.this.context_ = context.get();
+                        ExnConnector.this.notifyAll();
+                        log.info("Optimiser-controller connected to ActiveMQ");
+                    }
+                }
+            },
             List.of(
-                // asynchronous topics for sending out controller status
+                // Asynchronous topics for sending out controller status.
+                // Synchronous publishers are created dynamically, not added
+                // here.
                 amplMessagePublisher,
                 metricListPublisher,
-                appStatusPublisher,
-                // synchronous communication with SAL via exn-middleware
-                findSalNodeCandidates,
-                findBrokerNodeCandidates,
-                findBrokerNodeCandidatesMultiple,
-                defineCluster,
-                getCluster,
-                labelNodes,
-                deployCluster,
-                deployApplication,
-                scaleOut,
-                scaleIn,
-                deleteCluster),
+                appStatusPublisher),
             List.of(
                 new Consumer("solver_status", solver_status_channel,
                     new SolverStatusMessageHandler(), true, true),
@@ -537,6 +520,7 @@ public class ExnConnector {
      *  first.
      */
     public List<NodeCandidate> findNodeCandidates(List<Requirement> requirements, String appID) {
+        Context context = getContext();
         Map<String, Object> msg;
         try {
             msg = Map.of(
@@ -547,15 +531,20 @@ public class ExnConnector {
                 e);
             return null;
         }
-        Map<String, Object> response = findBrokerNodeCandidates.sendSync(msg, appID, null, false);
-        // Note: we do not call extractPayloadFromExnResponse here, since this
-        // response does not come from the exn-middleware, so will not be
-        // packaged into a string.
-        ObjectNode jsonBody = mapper.convertValue(response, ObjectNode.class);
-        // Note: If the result is empty, the body will be an empty object
-        // instead of an empty array, so we use `JsonNode.OverwriteMode.ALL`
-        List<JsonNode> result = Arrays.asList(mapper.convertValue(jsonBody.withArray("/body", JsonNode.OverwriteMode.ALL, true), JsonNode[].class));
-        result.sort((JsonNode c1, JsonNode c2) -> {
+        SyncedPublisher findBrokerNodeCandidates = new SyncedPublisher(
+            "findBrokerNodeCandidates" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.cfsb.get_node_candidates", true, true);
+        try {
+            context.registerPublisher(findBrokerNodeCandidates);
+            Map<String, Object> response = findBrokerNodeCandidates.sendSync(msg, appID, null, false);
+            // Note: we do not call extractPayloadFromExnResponse here, since this
+            // response does not come from the exn-middleware, so will not be
+            // packaged into a string.
+            ObjectNode jsonBody = mapper.convertValue(response, ObjectNode.class);
+            // Note: If the result is empty, the body will be an empty object
+            // instead of an empty array, so we use `JsonNode.OverwriteMode.ALL`
+            List<JsonNode> result = Arrays.asList(mapper.convertValue(jsonBody.withArray("/body", JsonNode.OverwriteMode.ALL, true), JsonNode[].class));
+            result.sort((JsonNode c1, JsonNode c2) -> {
                 long rank1 = c1.at("/rank").longValue();
                 long rank2 = c2.at("/rank").longValue();
                 double score1 = c1.at("/score").doubleValue();
@@ -574,12 +563,15 @@ public class ExnConnector {
                 else if (cpu1 != cpu2) return cpu1 - cpu2;
                 else return ram1 - ram2;
             });
-        return result.stream()
-            .map(candidate ->
-                mapper.convertValue(
-                    ((ObjectNode)candidate).deepCopy().remove(List.of("score", "rank")),
-                    NodeCandidate.class))
-            .collect(Collectors.toList());
+            return result.stream()
+                .map(candidate ->
+                    mapper.convertValue(
+                        ((ObjectNode)candidate).deepCopy().remove(List.of("score", "rank")),
+                        NodeCandidate.class))
+                .collect(Collectors.toList());
+        } finally {
+            context.unregisterPublisher(findBrokerNodeCandidates.key());
+        }
     }
 
     /**
@@ -602,6 +594,7 @@ public class ExnConnector {
         // This is an almost literal copy of `findNodeCandidates`; if there's
         // a third method with the same functionality I'll start thinking
         // about unifying them.
+        Context context = getContext();
         Map<String, Object> msg;
         try {
             msg = Map.of(
@@ -612,15 +605,20 @@ public class ExnConnector {
                 e);
             return null;
         }
-        Map<String, Object> response = findBrokerNodeCandidatesMultiple.sendSync(msg, appID, null, false);
-        // Note: we do not call extractPayloadFromExnResponse here, since this
-        // response does not come from the exn-middleware, so will not be
-        // packaged into a string.
-        ObjectNode jsonBody = mapper.convertValue(response, ObjectNode.class);
-        // Note: If the result is empty, the body will be an empty object
-        // instead of an empty array, so we use `JsonNode.OverwriteMode.ALL`
-        List<JsonNode> result = Arrays.asList(mapper.convertValue(jsonBody.withArray("/body", JsonNode.OverwriteMode.ALL, true), JsonNode[].class));
-        result.sort((JsonNode c1, JsonNode c2) -> {
+        SyncedPublisher findBrokerNodeCandidatesMultiple = new SyncedPublisher(
+            "findBrokerNodeCandidatesMultiple" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.cfsb.get_node_candidates_multi", true, true);
+        try {
+            context.registerPublisher(findBrokerNodeCandidatesMultiple);
+            Map<String, Object> response = findBrokerNodeCandidatesMultiple.sendSync(msg, appID, null, false);
+            // Note: we do not call extractPayloadFromExnResponse here, since this
+            // response does not come from the exn-middleware, so will not be
+            // packaged into a string.
+            ObjectNode jsonBody = mapper.convertValue(response, ObjectNode.class);
+            // Note: If the result is empty, the body will be an empty object
+            // instead of an empty array, so we use `JsonNode.OverwriteMode.ALL`
+            List<JsonNode> result = Arrays.asList(mapper.convertValue(jsonBody.withArray("/body", JsonNode.OverwriteMode.ALL, true), JsonNode[].class));
+            result.sort((JsonNode c1, JsonNode c2) -> {
                 long rank1 = c1.at("/rank").longValue();
                 long rank2 = c2.at("/rank").longValue();
                 double score1 = c1.at("/score").doubleValue();
@@ -639,31 +637,28 @@ public class ExnConnector {
                 else if (cpu1 != cpu2) return cpu1 - cpu2;
                 else return ram1 - ram2;
             });
-        return result.stream()
-            .map(candidate ->
-                mapper.convertValue(
-                    ((ObjectNode)candidate).deepCopy().remove(List.of("score", "rank")),
-                    NodeCandidate.class))
-            .collect(Collectors.toList());
+            return result.stream()
+                .map(candidate ->
+                    mapper.convertValue(
+                        ((ObjectNode)candidate).deepCopy().remove(List.of("score", "rank")),
+                        NodeCandidate.class))
+                .collect(Collectors.toList());
+        } finally {
+            context.unregisterPublisher(findBrokerNodeCandidatesMultiple.key());
+        }
     }
 
     /**
      * Get list of node candidates from the resource broker that fulfil the
-     * given requirements.
-     *
-     * <p>Note that we cannot convert the result to a list containing {@code
-     * org.ow2.proactive.sal.model.NodeCandidate} instances, since the broker
-     * adds the additional fields {@code score} and {@code ranking}.  Instead
-     * we return a JSON {@code ArrayNode} containing {@code ObjectNode}s in
-     * the format specified at
-     * https://github.com/ow2-proactive/scheduling-abstraction-layer/blob/master/documentation/nodecandidates-endpoints.md#71--filter-node-candidates-endpoint
-     * but with these two additional attributes.
+     * given requirements.  We sort the list so that "smaller" candidates
+     * (fewer cores, memory) come first.
      *
      * @param requirements The list of requirements.
      * @param appID The application ID.
      * @return A list containing node candidates, or null in case of error.
      */
     public List<NodeCandidate> findNodeCandidatesFromSal(List<Requirement> requirements, String appID) {
+        Context context = getContext();
         Map<String, Object> msg;
         try {
             msg = Map.of(
@@ -674,13 +669,19 @@ public class ExnConnector {
                 e);
             return null;
         }
-        Map<String, Object> response = findSalNodeCandidates.sendSync(msg, appID, null, false);
-        JsonNode payload = extractPayloadFromExnResponse(response, "findNodeCandidatesFromSal");
-        if (payload.isMissingNode()) return null;
-        if (!payload.isArray()) return null;
-        List<NodeCandidate> candidates = Arrays.asList(mapper.convertValue(payload, NodeCandidate[].class));
-        // We try to choose candidates with lower hardware requirements; sort by cores, ram
-        candidates.sort((NodeCandidate c1, NodeCandidate c2) -> {
+        SyncedPublisher findSalNodeCandidates = new SyncedPublisher(
+            "findSalNodeCandidates" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.exn.sal.nodecandidate.get",
+            true, true);
+        try {
+            context.registerPublisher(findSalNodeCandidates);
+	    Map<String, Object> response = findSalNodeCandidates.sendSync(msg, appID, null, false);
+            JsonNode payload = extractPayloadFromExnResponse(response, "findNodeCandidatesFromSal");
+            if (payload.isMissingNode()) return null;
+            if (!payload.isArray()) return null;
+            List<NodeCandidate> candidates = Arrays.asList(mapper.convertValue(payload, NodeCandidate[].class));
+            // We try to choose candidates with lower hardware requirements; sort by cores, ram
+            candidates.sort((NodeCandidate c1, NodeCandidate c2) -> {
                 int cpu1 = c1.getHardware().getCores();
                 int cpu2 = c2.getHardware().getCores();
                 long ram1 = c1.getHardware().getRam();
@@ -688,7 +689,10 @@ public class ExnConnector {
                 if (cpu1 != cpu2) return cpu1 - cpu2;
                 else return Math.toIntExact(ram1 - ram2);
             });
-        return candidates;
+            return candidates;
+        } finally {
+            context.unregisterPublisher(findSalNodeCandidates.key());
+        }
     }
 
     /**
@@ -732,6 +736,7 @@ public class ExnConnector {
     public boolean defineCluster(String appID, String clusterName, ObjectNode cluster) {
         // https://openproject.nebulouscloud.eu/projects/nebulous-collaboration-hub/wiki/deployment-manager-sal-1#specification-of-endpoints-being-developed
         Main.logFile("define-cluster-" + appID + ".json", cluster.toPrettyString());
+        Context context = getContext();
         Map<String, Object> msg;
         try {
             msg = Map.of("metaData", Map.of("user", "admin"),
@@ -741,9 +746,17 @@ public class ExnConnector {
                 e);
             return false;
         }
-        Map<String, Object> response = defineCluster.sendSync(msg, appID, null, false);
-        JsonNode payload = extractPayloadFromExnResponse(response, "defineCluster");
-        return payload.asBoolean();
+        SyncedPublisher defineCluster = new SyncedPublisher(
+            "defineCluster" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.exn.sal.cluster.define", true, true);
+        try {
+            context.registerPublisher(defineCluster);
+            Map<String, Object> response = defineCluster.sendSync(msg, appID, null, false);
+            JsonNode payload = extractPayloadFromExnResponse(response, "defineCluster");
+            return payload.asBoolean();
+        } finally {
+            context.unregisterPublisher(defineCluster.key());
+        }
     }
 
     /**
@@ -754,10 +767,19 @@ public class ExnConnector {
      * @return The cluster definition, or null in case of error.
      */
     public JsonNode getCluster(String appID, String clusterName) {
+        Context context = getContext();
         Map<String, Object> msg = Map.of("metaData", Map.of("user", "admin", "clusterName", clusterName));
-        Map<String, Object> response = getCluster.sendSync(msg, appID, null, false);
-        JsonNode payload = extractPayloadFromExnResponse(response, "getCluster");
-        return payload.isMissingNode() ? null : payload;
+        SyncedPublisher getCluster = new SyncedPublisher(
+            "getCluster" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.exn.sal.cluster.get", true, true);
+        try {
+            context.registerPublisher(getCluster);
+	    Map<String, Object> response = getCluster.sendSync(msg, appID, null, false);
+            JsonNode payload = extractPayloadFromExnResponse(response, "getCluster");
+            return payload.isMissingNode() ? null : payload;
+        } finally {
+            context.unregisterPublisher(getCluster.key());
+        }
     }
 
     /**
@@ -768,6 +790,7 @@ public class ExnConnector {
      * @param labels A map from node name to label.
      */
     public boolean labelNodes(String appID, String clusterID, JsonNode labels) {
+        Context context = getContext();
         Map<String, Object> msg;
         try {
             msg = Map.of("metaData", Map.of("user", "admin", "clusterName", clusterID),
@@ -777,9 +800,17 @@ public class ExnConnector {
                 e);
             return false;
         }
-        Map<String, Object> response = labelNodes.sendSync(msg, appID, null, false);
-        JsonNode payload = extractPayloadFromExnResponse(response, "labelNodes");
-        return payload.isMissingNode() ? false : true;
+        SyncedPublisher labelNodes = new SyncedPublisher(
+            "labelNodes" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.exn.sal.cluster.label", true, true);
+        try {
+            context.registerPublisher(labelNodes);
+	    Map<String, Object> response = labelNodes.sendSync(msg, appID, null, false);
+            JsonNode payload = extractPayloadFromExnResponse(response, "labelNodes");
+            return payload.isMissingNode() ? false : true;
+        } finally {
+            context.unregisterPublisher(labelNodes.key());
+        }
     }
 
     /**
@@ -794,11 +825,20 @@ public class ExnConnector {
      */
     public boolean deployCluster(String appID, String clusterName) {
         // https://openproject.nebulouscloud.eu/projects/nebulous-collaboration-hub/wiki/deployment-manager-sal-1#specification-of-endpoints-being-developed
+        Context context = getContext();
         Map<String, Object> msg = Map.of("metaData",
             Map.of("user", "admin", "clusterName", clusterName));
-        Map<String, Object> response = deployCluster.sendSync(msg, appID, null, false);
-        JsonNode payload = extractPayloadFromExnResponse(response, "deployCluster");
-        return payload.asBoolean();
+        SyncedPublisher deployCluster = new SyncedPublisher(
+            "deployCluster" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.exn.sal.cluster.deploy", true, true);
+        try {
+            context.registerPublisher(deployCluster);
+	    Map<String, Object> response = deployCluster.sendSync(msg, appID, null, false);
+            JsonNode payload = extractPayloadFromExnResponse(response, "deployCluster");
+            return payload.asBoolean();
+        } finally {
+            context.unregisterPublisher(deployCluster.key());
+        }
     }
 
     /**
@@ -820,6 +860,7 @@ public class ExnConnector {
             .put("flags", "");
         Main.logFile("deploy-application-" + appID + ".json", body.toPrettyString());
         Main.logFile("deploy-application-" + appID + ".yaml", kubevela);
+        Context context = getContext();
         Map<String, Object> msg;
         try {
             String bodyStr = mapper.writeValueAsString(body);
@@ -829,9 +870,17 @@ public class ExnConnector {
             log.error("Could not convert JSON to string (this should never happen)", e);
             return -1;
         }
-        Map<String, Object> response = deployApplication.sendSync(msg, appID, null, false);
-        JsonNode payload = extractPayloadFromExnResponse(response, "deployApplication");
-        return payload.asLong();
+        SyncedPublisher deployApplication = new SyncedPublisher(
+            "deployApplication" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.exn.sal.cluster.deployapplication", true, true);
+        try {
+            context.registerPublisher(deployApplication);
+            Map<String, Object> response = deployApplication.sendSync(msg, appID, null, false);
+            JsonNode payload = extractPayloadFromExnResponse(response, "deployApplication");
+            return payload.asLong();
+        } finally {
+            context.unregisterPublisher(deployApplication.key());
+        }
     }
 
     /**
@@ -845,6 +894,7 @@ public class ExnConnector {
      * @param nodesToAdd The additional nodes to add.
      */
     public void scaleOut(String appID, String clusterName, ArrayNode nodesToAdd) {
+        Context context = getContext();
         Map<String, Object> msg;
         try {
             msg = Map.of("metaData", Map.of("user", "admin",
@@ -854,12 +904,20 @@ public class ExnConnector {
             log.error("Could not convert JSON to string (this should never happen)", e);
             return;
         }
-        Map<String, Object> response = scaleOut.sendSync(msg, appID, null, false);
-        // Called for side-effect only; we want to log errors.  The return
-        // value from scaleOut is the same as getCluster, but since we have to
-        // poll for cluster status anyway to make sure the new machines are
-        // running, we do not return it here.
-        JsonNode payload = extractPayloadFromExnResponse(response, "scaleOut");
+        SyncedPublisher scaleOut = new SyncedPublisher(
+            "scaleOut" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.exn.sal.cluster.scaleout", true, true);
+        try {
+            context.registerPublisher(scaleOut);
+            Map<String, Object> response = scaleOut.sendSync(msg, appID, null, false);
+            // Called for side-effect only; we want to log errors.  The return
+            // value from scaleOut is the same as getCluster, but since we have to
+            // poll for cluster status anyway to make sure the new machines are
+            // running, we do not return it here.
+            JsonNode payload = extractPayloadFromExnResponse(response, "scaleOut");
+        } finally {
+            context.unregisterPublisher(scaleOut.key());
+        }
     }
 
     /**
@@ -872,6 +930,7 @@ public class ExnConnector {
      */
     public boolean scaleIn(String appID, String clusterName, List<String> superfluousNodes) {
         // https://openproject.nebulouscloud.eu/projects/nebulous-collaboration-hub/wiki/deployment-manager-sal-1#specification-of-endpoints-being-developed
+        Context context = getContext();
         ArrayNode body = mapper.createArrayNode();
         superfluousNodes.forEach(nodeName -> body.add(nodeName));
         Map<String, Object> msg;
@@ -882,9 +941,17 @@ public class ExnConnector {
             log.error("Could not convert JSON to string (this should never happen)", e);
             return false;
         }
-        Map<String, Object> response = scaleIn.sendSync(msg, appID, null, false);
-        JsonNode payload = extractPayloadFromExnResponse(response, "scaleIn");
-        return payload.asBoolean();
+        SyncedPublisher scaleIn = new SyncedPublisher(
+            "scaleIn" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.exn.sal.cluster.scalein", true, true);
+        try {
+            context.registerPublisher(scaleIn);
+            Map<String, Object> response = scaleIn.sendSync(msg, appID, null, false);
+            JsonNode payload = extractPayloadFromExnResponse(response, "scaleIn");
+            return payload.asBoolean();
+        } finally {
+            context.unregisterPublisher(scaleIn.key());
+        }
     }
 
     /**
@@ -896,11 +963,20 @@ public class ExnConnector {
      */
     public boolean deleteCluster(String appID, String clusterName) {
         // https://openproject.nebulouscloud.eu/projects/nebulous-collaboration-hub/wiki/deployment-manager-sal-1#specification-of-endpoints-being-developed
+        Context context = getContext();
         Map<String, Object> msg = Map.of("metaData",
             Map.of("user", "admin", "clusterName", clusterName));
-        Map<String, Object> response = deleteCluster.sendSync(msg, appID, null, false);
-        JsonNode payload = extractPayloadFromExnResponse(response, "deleteCluster");
-        return payload.asBoolean();
+        SyncedPublisher deleteCluster = new SyncedPublisher(
+            "deleteCluster" + publisherNameCounter.incrementAndGet(),
+            "eu.nebulouscloud.exn.sal.cluster.delete", true, true);
+        try {
+            context.registerPublisher(deleteCluster);
+            Map<String, Object> response = deleteCluster.sendSync(msg, appID, null, false);
+            JsonNode payload = extractPayloadFromExnResponse(response, "deleteCluster");
+            return payload.asBoolean();
+        } finally {
+            context.unregisterPublisher(deleteCluster.key());
+        }
     }
 
     // ----------------------------------------

--- a/optimiser-controller/src/main/java/eu/nebulouscloud/optimiser/controller/Main.java
+++ b/optimiser-controller/src/main/java/eu/nebulouscloud/optimiser/controller/Main.java
@@ -168,12 +168,7 @@ public class Main implements Callable<Integer> {
                 activemq_host, activemq_port);
             activeMQConnector
               = new ExnConnector(activemq_host, activemq_port,
-                  activemq_user, activemq_password,
-                  new ConnectorHandler() {
-                    public void onReady(AtomicReference<Context> context) {
-                      log.info("Optimiser-controller connected to ActiveMQ");
-                    }
-                  });
+                  activemq_user, activemq_password);
         } else {
             log.debug("ActiveMQ login info not set, only operating locally.");
         }

